experimental/dds/tree2/src/shared-tree/undoRedoHandler.ts

```diff
@@ -0,0 +1,65 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IRevertible, UndoRedoStackManager } from "@fluidframework/undo-redo";
+import { UndoRedoManagerCommitType } from "../core";
+import { ISharedTreeView } from ".";
+
+/**
+ * A shared tree view undo redo handler that will add all local undoable tree changes to the provided
+ * undo redo stack manager.
+ *
+ * @alpha
+ */
+export class SharedTreeViewUndoRedoHandler {
+	private detach?: () => void;
+
+	public constructor(private readonly stackManager: UndoRedoStackManager) {}
+
+	public attachTree(tree: ISharedTreeView) {
+		this.detach = tree.events.on("undoable", this.treeDeltaHandler);
+	}
+	public detachTree(tree: ISharedTreeView) {
+		this.detach?.();
+	}
+
+	private readonly treeDeltaHandler = (
+		type: UndoRedoManagerCommitType,
+		target: ISharedTreeView,
+	) => {
+		this.stackManager.pushToCurrentOperation(new SharedTreeViewRevertible(type, target));
+	};
+}
+
+/**
+ * Provides an interface for reverting a change to a SharedTree. SharedTree manages its own undo stack so this
+ * revertible stores no information about the commit being reverted other than whether it needs to be an undo or redo.
+ *
+ * @alpha
+ */
+export class SharedTreeViewRevertible implements IRevertible {
```

Comment:
Is this only used by `SharedTreeViewUndoRedoHandler`? Does it need to be exported?

Having a class with these two fields might be a bit overkill. Consider doing something like:

```
function getRevertible(view: ISharedTreeView, type: "undo" | "redo"): IRevertible {
	return {
		revert: type === "undo" ? () => view.undo() : () => view.redo(),
		discard: () => {},
	};
}
```

The `IRevertible` itself does not actually need any logic in it; everything that it will ever do is predetermined once it has been created. If we were micro-opping this (don't :D), we could actually just create two static objects per view, one `IRevertible` for undo and one `IRevertible` for redo.